pkg/util/hostname.go

```diff
@@ -122,9 +137,14 @@ func GetHostname() (string, error) {
 	configName := config.Datadog.GetString("hostname")
 	err = ValidHostname(configName)
 	if err == nil {
-		cache.Cache.Set(cacheHostnameKey, configName, cache.NoExpiration)
-		setHostnameProvider("configuration")
-		return configName, err
+		provider = HostnameProviderConfiguration
+		hostnameData := HostnameData{Hostname: configName, Provider: provider}
+		cache.Cache.Set(cacheHostnameKey, hostnameData, cache.NoExpiration)
+		setHostnameProvider(provider)
```

Comment:
this _looks like_ quite some copy-paste of code to cache the hostname, hostname provider, and set the hostname provider. It could all be extracted out in one function no? Something like `func setHostnameData(name, provider)` that'd take care of everything.

For example, I don't think there's ever a case where we want to update the `HostnameData` that's in the cache but not update the value that's stored in the expvar. Please check though.

Reply:
Code factorize.

I did not change at https://github.com/DataDog/datadog-agent/pull/4504/files/4f1229fd8cbbdeaeb7117d74adaa0acbe6414574#diff-1d9d99d196299ff9a78c3e928058494aL139 because it is not clear if `setHostnameProvider("")` is equivalent to not calling `setHostnameProvider`.

In addition, I cannot update `setHostnameData` to 
```go
func setHostnameData(cacheHostnameKey string, hostname string, provider string) HostnameData {
	hostnameData := HostnameData{Hostname: hostname, Provider: provider}
	cache.Cache.Set(cacheHostnameKey, hostnameData, cache.NoExpiration)
	if provider != "" {  // *** CODE UPDATED ***
             setHostnameProvider(provider)
        }
	return hostnameData
}
```

because `provider` can be empty at https://github.com/DataDog/datadog-agent/pull/4504/commits/ff0055c86891e5bec4b866f912b336dfffce75cd#diff-1d9d99d196299ff9a78c3e928058494aR276.

Reply:
in which cases can the provider be empty? I can't think of a case where the hostname would be non-empty *and* the provider would be empty. Am I missing something?

Reply:
In `GetHostnameData`, if no `if` matches, `provider` is empty. I do not know if it can happen in practice.

Reply:
discussed offline: when `hostname` is non-empty, the `provider` is non-empty as well, so we can factorize the logic fine 👍